tools/tester.nim

```diff
@@ -265,39 +328,59 @@ if file.len == 0:
           stdout.write(p.output)
           quit(1)
 
-        dirs.add (it.path, exe)
+        dirs.add (it.path, RunnerDesc(exe: exe, args: @["${args}", "${file}"]))
+      elif fileExists(it.path / "runner.txt"):
+        # an external runner is used for the directory
+        let cmdFile = it.path / "runner.txt"
+        stdout.write("[Parsing] ")
+        stdout.writeLine(cmdFile)
+
+        let
+          f = open(cmdFile, fmRead)
+          line = f.readLine()
+        f.close()
+
+        let desc = parseDesc(line)
+        if desc.isSome:
+          dirs.add (it.path, desc.unsafeGet)
+        else:
+          stdout.writeLine("Failure" + fgRed)
+          quit(1)
 
   stdout.flushFile()
 
   var
     total = 0
     success = 0
   # now run the tests:
-  # XXX: concurrent execution of the processes is currently not possible,
-  #      as they'd clobber each others' output
-  for (dir, exe) in dirs.items:
+  # XXX: parallel execution of tests is still missing
+  for (dir, runner) in dirs.items:
     for it in walkDir(dir, relative=false):
       if it.path.endsWith(".test"):
         inc total
-        let p =
-          startProcess(getAppFilename(),
-                       args=["--runner:" & exe, it.path.relativePath(currDir)],
-                       options={poParentStreams})
-        if p.waitForExit() == 0:
+        if runTest(runner, it.path.relativePath(currDir)):
           inc success
-        p.close()
 
   if success == total:
     echo "Success! Ran ", success, " tests"
   else:
     echo "Failure"
     quit(1)
 else:
+  # legacy mode:
+  var desc = RunnerDesc(exe: runner)
```

Comment:
IIUC a `var` is not required.

Reply:
Yep, it's not, and the `RunnerDesc` is also incomplete (it misses the argument specification).